File: src/main/java/chat/teco/tecochat/comment/presentation/CommentController.java
package chat.teco.tecochat.comment.presentation;

import static org.springframework.http.ResponseEntity.created;
import static org.springframework.http.ResponseEntity.ok;

import chat.teco.tecochat.auth.Auth;
import chat.teco.tecochat.comment.application.CommentService;
import chat.teco.tecochat.comment.application.dto.DeleteCommentCommand;
import chat.teco.tecochat.comment.application.dto.UpdateCommentCommand;
import chat.teco.tecochat.comment.application.dto.WriteCommentCommand;
import chat.teco.tecochat.comment.presentation.request.UpdateCommentRequest;
import chat.teco.tecochat.comment.presentation.request.WriteCommentRequest;
import chat.teco.tecochat.comment.presentation.response.CreatedIdResponse;
import chat.teco.tecochat.common.util.UriUtil;
import java.net.URI;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RequiredArgsConstructor
@RestController
@RequestMapping("/comments")
public class CommentController {

    private final CommentService commentService;

    @PostMapping
    ResponseEntity<CreatedIdResponse> write(
            @Auth Long memberId,
            @RequestBody WriteCommentRequest request
    ) {
        Long id = commentService.write(
                new WriteCommentCommand(request.chatId(), memberId, request.content())
        );
        URI uri = UriUtil.buildURI("/{id}", id);
        return created(uri).body(new CreatedIdResponse(id));
    }

    @PatchMapping("/{id}")
    ResponseEntity<Void> update(
            @PathVariable("id") Long commentId,
            @Auth Long memberId,
            @RequestBody UpdateCommentRequest request) {
        commentService.update(
                new UpdateCommentCommand(commentId, memberId, request.content())
        );
        return ok().build();
    }

    @DeleteMapping("/{id}")
    ResponseEntity<Void> delete(
            @PathVariable("id") Long commentId,
            @Auth Long memberId
    ) {
        commentService.delete(
                new DeleteCommentCommand(commentId, memberId)
        );
        return ok().build();
    }
}

File: src/main/kotlin/chat/teco/tecochat/ui/CommentController.kt
package chat.teco.tecochat.ui

import chat.teco.tecochat.application.CommentResponse
import chat.teco.tecochat.application.CommentService
import chat.teco.tecochat.application.UpdateCommentRequest
import chat.teco.tecochat.application.WriteCommentRequest
import chat.teco.tecochat.application.WriteCommentResponse
import chat.teco.tecochat.security.Auth
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PatchMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import java.net.URI

@RestController
@RequestMapping("/comments")
class CommentController(
    private val commentService: CommentService,
) {

    @PostMapping
    fun write(
        @Auth memberId: Long,
        @RequestBody request: WriteCommentRequest,
    ): ResponseEntity<WriteCommentResponse> {
        val id = commentService.write(memberId, request)
        val uri = URI(id.toString())
        return ResponseEntity.created(uri).body(WriteCommentResponse(id))
    }

    @PatchMapping("/{id}")
    fun update(
        @Auth memberId: Long,
        @PathVariable("id") commentId: Long,
        @RequestBody request: UpdateCommentRequest,
    ): ResponseEntity<Void> {
        commentService.update(memberId, commentId, request)
        return ResponseEntity.ok().build()
    }

    @DeleteMapping("/{id}")
    fun delete(
        @Auth memberId: Long,
        @PathVariable("id") commentId: Long,
    ): ResponseEntity<Void> {
        commentService.delete(memberId, commentId)
        return ResponseEntity.ok().build()
    }

    @GetMapping
    fun findAllByChatId(
        @RequestParam("chatId") chatId: Long,
    ): ResponseEntity<List<CommentResponse>> {
        return ResponseEntity.ok(commentService.findAllByChatId(chatId))
    }
}

